### ISSUE TITLE
[feature] Add nginx filters and bad-URI filters?

### ISSUE THREAD

**Issue by cerlestes** [fail2ban/fail2ban#2583]
I've been using fail2ban for years now, but still the official distribution contains no nginx filters. I was wondering: is there a reason for that?

For all my server setups (20+ production servers for various projects with currently 100k-1M req/day total) I've been using the following two filters since around 2012 with great success.

https://gist.github.com/cerlestes/68634e2d808fea2a33ddc7b1bd5fb5da
Copy of apache-badbots with nginx datepattern.

https://gist.github.com/cerlestes/1d6f1549f06350f7c4f4
Catches a lot of file and directory names that should not be accessed under normal conditions. Ignored when server returns a good HTTP status code. This catches malicious requests like this:
```
51.255.162.143 - - [29/Dec/2019:21:13:29 +0100] "GET / HTTP/1.1" 301 178 "-" "Go-http-client/1.1"
51.255.162.143 - - [29/Dec/2019:21:13:30 +0100] "GET /install/lib/ajaxHandlers/ajaxServerSettingsChk.php?rootUname=;echo -n HellorConfig|md5sum  HTTP/1.1" 404 178 "-" "-"
```

Do y'all want me to add them via pull-request? I'd also add a nginx-common.conf that has the datepatterns for access and error logs.

**Comment by sebres** [comment]
> I was wondering: is there a reason for that?

Few of possible reasons are mentioned in the article [Best-practice](https://github.com/fail2ban/fail2ban/wiki/Best-practice) of our wiki - you can do the "pre-filtering" with the aid of nginx features more effectively - for example some zone in 40x location can capture several wrong or invalid attempts and can log it only once (or at least only the evildoers without any parasitic traffic in log), so fail2ban should not search the whole access log for a matches with some doubtful filters like `apache-badbots` (yes - it's part of fail2ban, and no - I don't think it is good idea to use that, at least with default settings).

> Do y'all want me to add them via pull-request?

The PR's are always welcome... Just some of that may not always make sense.

**Comment by cerlestes** [comment]
Oh okay, I understand that. I didn't know that was considered the idomatic approach. May I ask why you dislike those filters though? Is there any reason apart from Python (fail2ban) being slower than C (nginx)? I've never had any problems with performance on any of my installations, so it never came to my mind to offload some of fail2ban's work to nginx. I've had fail2ban search hundreds of megabytes of logfiles without problems.

If you're telling me that you don't want filters like `apache-badbots`, then I won't create a PR for the nginx variants and just stick to my gist files for myself, that's fine.

But IMHO I really like the fact that I don't have to change anything about the webserver configuration for fail2ban to catch bots and vulnerability scanners. It's a completely separate addon that way, I think that should be one of the goals of fail2ban, and it indeed seemed to be from my experience with the software so far. What do you think about that?

**Comment by sebres** [comment]
> Is there any reason apart from Python (fail2ban) being slower than C (nginx)?

Well, it's not about slower/faster, moreover nginx (or either the admin using/configuring it) knows better which requests are "broken" or invalid by design and which going "broken" by accident.

Note I don't suggest to implement badbots in nginx, I'm against any form of black-listing like that.
For example if you'd configure a location logging all 30x-50x requests into separate log, you'd be able to ban 99.9% of all the bots without filtering that over very-long black-listing rules.

> example: I have valid users that use old bookmarks or incorrect API and constantly cause 404 errors;

1. But he will surely not do that continuously? 
So a large `maxretry` / short `findrime` combination would be enough either, or even usage of incremental ban (in v.0.11) with initial short `bantime` may help here too.
Also it is possible to use zones for that (e. g. limit_req for some long enough time in such location to ignore legitimate users).
2. You would probably know all this locations, so why you'd configure nginx to log them into "intruder" log file monitored by fail2ban?
3. Many of that URIs can be (individually) combined to much shorter URIs, so if I don't have perl/php on the machine (or some locations of web-server), my black-listing filter (for this kind of requests) looks simply like `(?:.php|.pl)$`.
Just compare it with all that or-ed combination in the "default" filters that are written without this "knowledge" about the machine.

The main point is - you alone knows which URIs are valid (or may be valid) in which locations, and vice versa which URIs are or may be definitely invalid.

> If you're telling me that you don't want filters like apache-badbots

Nope, I had no such intention... you can make a PR (and it will be even accepted, if it'd look good).
I wanted just point you (and other people reading that) that it is not necessarily advisable to use such filters per default (without some customization).

Also note there are another open PRs that may be similar yours - [is:pr is:open nginx](https://github.com/fail2ban/fail2ban/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+nginx)

**Comment by cerlestes** [comment]
I see, thank you for your explanation! But I have to politely disagree with the view that the nginx admin (or even the application developer) will always know what routes are valid or invalid, or even be able to update those values in a manner that is better than having fail2ban filter it without any pre-filtering. 

In some of my projects we've had the possibility to write explicit whitelist nginx configurations which contained all known routes, and this was amazing in those projects. But that's simply not possible for all projects, and neither is defining which routes are invalid (expect for some well-known malicious strings like I'm trying to do here). Think of a load-balancer setup for a very diverse and dynamic backend landscape. Or a generic webhoster kind of scenario. Then it would be very hard for nginx to be configured in a way that allowed for it to decide which requests are valid of invalid - it needs to ask the backend whether it is valid or not. And as you've said yourself, some errors are by design and should not go into a `please_fail2ban.log` file to avoid having those users banned simply because they caused error status codes.

This is exactly where I see the prosed filter being used (and again, I've been using this for years now without any false positive to my best knowledge, but greatly reduced bot/malicious traffic): setups that don't have special webserver/loadbalancer/proxy configurations, but still want bots to be banned as soon as possible from their systems without a lot of customization. I think this fits probably around 90% of the fail2ban installations, who are just rolling with the default configs for a bit of extra security.

I'll see if I can put a good pull-request or two together. Of course I won't half-ass it, no worries. :)
I'll make sure to include explanation about the best practices you mentioned.

Also thanks for pointing out the nginx pull-requests, I was initially only looking for existing issues but not pull-requests.

**Comment by sebres** [comment]
> But that's simply not possible for all projects

Hmm... why? or what exactly is not possible?
If you've a location serving static URIs only, you can **completely** prohibit all URIs like php/pl/cgi/whatever without large black-listing set of rules.
If you've a location serving dynamic URIs, you know which are valid and can **completely** prohibit all not valid (e. g. all 404) from there.
Many of (valid) requests tested by the bots normally requiring an authenticated user, so you can **completely** prohibit all requests for users that are not authenticated (or even all 401/403 attempts).
Etc pp.

> Think of a load-balancer setup for a very diverse and dynamic backend landscape. 

Well, to have a load balancer in order to speedup the server and at same time to introduce a "handbrake" like such filters in fail2ban (e. g. very expensive washing out the system cache), that due to common blacklisting would even not capture **all** of the failures in a "very diverse and dynamic backend landscape" - is simply worse strategy in my opinion.

But it is also a bit matter of taste.

**Comment by cerlestes** [comment]
> Hmm... why? or what exactly is not possible?

I have multiple nginx setups where I'm not only unknowing of what routes are valid or invalid, but I have no knowledge about the locations or their properties at all, aside from it existing. This is basically always the case with a simple nginx-based reverse-proxy load balancer. Only in nginx configs that are tightly bound to a specific application I can know what routes see which status codes as abnormal, and specify optimizations for these routes (like cache headers).

> or even all 401/403 attempts

I agree those are easily handable, but I'm talking about 404 errors specifically. Bots like OpenVAS, Nessus and script-kiddie scanners looking for files that don't exist but would be lucrative targets to find, like administrative interfaces, databases, hidden directories, ...

> to have a load balancer in order to speedup the server and at same time to introduce a "handbrake"

Speed-up is a by-product of a good load balancer, but I'd argue that the main reason for having a load balancer is TLS termination, spam blocking and reducing the load on the backend servers (e.g. by HTTP pipelining, queuing, CDN, ...) - at least it was like that in all of my projects so far. It was never about speeding up anything, it was about shielding the backend from raw HTTP traffic. And fail2ban absolutely falls within that range IMO.

> But it is also a bit matter of taste.

Yes, absolutely. :) Again, I just feel like fail2ban should have these filters for the people who need them, and I consider it a good place to implement them. Even if there are surely better places to handle this kind of filtering whenever possible.

**Comment by sebres** [comment]
> but I'm talking about 404 errors specifically. 

So if tomorrow some group of script kiddies starting to search for some new "popular" URLs (like \*coin, new CRM, new service, etc) you'd extend your "bad" URIs list?
And what would you do if they starting to search for existence of fewer lucrative targets, for example any kind of URIs like `aaa...` to `zzz...`?

> like administrative interfaces, databases, hidden directories, ...

And exactly this services expecting some privileges (so as written above you can distinguish between authenticated user and not authenticated user).
Moreover it is sometimes enough to place that all into some different common location like `/admin/...` to avoid matching of 90% of all possible attempts as well as to make the monitoring fewer expensive (like N of 404th attempts in M minutes **in this location** cause a ban no matter what they are).

> It's not slowing down traffic in transit

Note that the server resources are finite (system and CPU cache, memory, CPU cores. etc)... as well as other things like file locks, etc.
So indirectly it does that.

**Comment by cerlestes** [comment]
> So if tomorrow some group of script kiddies starting to search for some new "popular" URLs (like *coin, new CRM, new service, etc) you'd extend your "bad" URIs list?

Yes, that's the idea. It filters for generic stuff, but also for well-known names. For example `wp-admin` is what 90% of the bots I see in log files query on first try (or second, if they query root first), meaning that alone bans 90% of malicious actors.

> And what would you do if they starting to search for existence of fewer lucrative targets, for example any kind of URIs like aaa... to zzz...?

No, I'd consider that out of scope of this filter. I've also never seen such an attack happen. IMO it would be a case for standard rate limiting rather than this filter.
